compiler/src/iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp

```diff
@@ -1192,11 +1192,35 @@ struct ElideRedundantTransfer : public OpRewritePattern<TensorTransferOp> {
   }
 };
 
+// Attempts to identify trivial case of chained transfer ops (A -> B -> C) and rewrite it as (A -> C)
+// Writes it as A -> B and A -> C relying on dead code elimination to remove the unused A -> B transfer.
+struct ElideChainedTransfer : public OpRewritePattern<TensorTransferOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(TensorTransferOp currTransferOp,
+                                PatternRewriter &rewriter) const override {
+    auto baseValue =
```

Comment:
This is not safe with barriers - a flow.tensor.barrier is supposed to force movement, whereas this will walk through those and find the origin. You can only really safely do a non-tied base lookup here - if *explicitly* a transfer > transfer > transfer you can elide, but any intervening ops has to break it.

If you find common ops breaking the chain (reshapes/etc) then we can add patterns to bubble those up as we do with others to try to get them out (so reshape > transfer > reshape > transfer > reshape > transfer could become reshape > transfer > transfer > transfer, which then you can fold).